datamapplot/static/js/table_of_contents.js

```diff
@@ -0,0 +1,186 @@
+// Helper to get the next nested list out of the table of contents.
+var getNextSibling = function (elem, selector) {
+    // Get the next sibling element
+    var sibling = elem.nextElementSibling;
+
+    // If there's no selector, return the first sibling
+    if (!selector) return sibling;
+
+    // If the sibling matches our selector, use it
+    // If not, jump to the next sibling and continue the loop
+    while (sibling) {
+        if (sibling.matches(selector)) return sibling;
+        sibling = sibling.nextElementSibling
+    }
+};
+
+function formatTocButtonHtml(icon, labelId) {
+    return `<button class="toc-btn" data-label-id="${labelId}">${icon}</button>`;
+}
+
+class TableOfContents {
+    constructor(
+        tocContainer,
+        datamap,
+        buttons,
+        icon,
+    ) {
+        this.container = tocContainer;
+        this.datamap = datamap;
+        this.elements = datamap.labelData;
+        this.rootLayerNo = Math.max(...datamap.labelData.map(e => e.layer_no));
+        this.parentChildMap = this.buildParentChildMap();
+        
+        this.container.innerHTML = `
+            <button class="expand-all-btn" data-expanded="false">Expand All</button>
+            ${this.buildTreeHtml(buttons, icon)}
+        `;
+    
+        this.setupCaretHandlers();
+        this.setupLabelHandlers(datamap);
+        this.setupExpandAllHandler();
+        this.highlightElements(this.elements);
```

Comment:
Create caching which we'll instantiate and re-use later...

```suggestion
        this.spanCache = new Map();
        this.parentChainCache = new Map();
        this.setupCaretHandlers();
        this.setupLabelHandlers(datamap);
        this.setupExpandAllHandler();
        this.initializeSpanCache();
        this.initializeParentChainCache();
        this.highlightElements(this.elements);
```